### PR TITLE
[Hashes]: Implementing hashes in rust

### LINKED ISSUE #2982: [Hashes]: Move Hashes to Rust FFI
**Is your feature request related to a problem? Please describe.**

Currently all of our hashing functionality is implemented in C++ and uses the TrezorCrypto library. The idea is eventually to reduce our dependencies and have them recoded in Rust.

**Describe the solution you'd like**

We will use [RustCrypto](https://github.com/RustCrypto) the library offers benefits such as not using dependencies which is crucial for WC

**Describe alternatives you've considered**

Parity crypto library

**Checklist**

- [ ] SHA
    - [ ] SHA1
    - [ ] SHA2
    - [ ] SHA3
- [ ] Keccak
- [ ] Ripemd
- [ ] Blake2
- [ ] Groestl

**Resources**

[Hashes](https://github.com/RustCrypto/hashes)

### PR DESCRIPTION
**Pull request by Milerius**
## Description

fix #2982 

## How to test

Run C++ unit tests

### CODE REVIEW

**Review by hewigovens — COMMENTED** [review]
Overall looks good to me, good to add `Rust` namespace 👍

1. can we use just `hash` as folder name? other name like `encoding` is not plural under `src` folder
2. `hash_wrapper` -> `hasher`
3. There are a lot of boilerplate code, maybe we can use macro or something else to simply it

```cpp
Data result(&raw_res.data[0], &raw_res.data[raw_res.size]);
std::free(raw_res.data);
```

File: tests/common/rust/bindgen/WalletCoreRsTests.cpp [hunk 1]
```diff
@@ -9,7 +9,7 @@
 
 TEST(RustBindgen, MoveParseFunctionArgument) {
     std::string arg = "10000000";
-    auto* result = parse_function_argument_to_bcs(arg.c_str());
+    auto* result = TW::Rust::parse_function_argument_to_bcs(arg.c_str());
```
Comment (hewigovens) [inline tests/common/rust/bindgen/WalletCoreRsTests.cpp:12]:
```suggestion
    auto* result = Rust::parse_function_argument_to_bcs(arg.c_str());
```

File: tests/common/rust/bindgen/WalletCoreRsTests.cpp [hunk 2]
```diff
@@ -9,7 +9,7 @@
...
     ASSERT_EQ(std::string(result), "8096980000000000");
-    free_string(result);
+    TW::Rust::free_string(result);
```
Comment (hewigovens) [inline tests/common/rust/bindgen/WalletCoreRsTests.cpp:14]:
```suggestion
    Rust::free_string(result);
```

**Review by hewigovens — APPROVED** [review]

**Review by rsrbk — APPROVED** [review]
Quite cool that we can write tests in the same file with the implementation

**Pull request merged**